### DISCUSSION TITLE
How can I use request body parser directly?

### DISCUSSION THREAD

**Discussion by HansLjy** [vercel/next.js#44212] [Help] [answered]
Here is the situation: I am currently implementing a backend api that handles file upload. Naturally I would like to use form-data instead of json as the request body so I need to disable body parser for this api path. However, I still need the body parser for other request method (GET, for example) at this api path. So how can I use the body parser of next directly? Or is there any way I can disable body parser for only some request method of a specific api path?

**Comment by icyJoseph** [comment, marked as answer]
MMm what do you think of something like:

```ts
// Next.js API route support: https://nextjs.org/docs/api-routes/introduction
import type { NextApiRequest, NextApiResponse } from "next";

export const config = {
  api: {
    bodyParser: false,
  },
};

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  const readable = req.read();

  const buffer = Buffer.from(readable);

  // not sure if there's any other checks to be made here,
  // I guess if it can be parsed, so be it?
  if (req.headers["content-type"] === "application/json") {
    try {
      const data = JSON.parse(buffer.toString());
      return res.status(200).json(data);
    } catch {
      return res.status(400).end();
    }
  }

  if (req.headers["content-type"]?.startsWith("multipart/form-data")) {
    console.log(buffer.toString());
    // handle buffer.toString() as you wish or
    // parse it into data manually or with a library
  }

  return res.status(404).end();
}

```

**Reply by nourishnew** [reply]
req.read() is not a function

**Reply by icyJoseph** [reply]
Are you using the App Router? This answer was for Pages Router.

**Reply by gcb** [reply]
where can i find information for app router?

Also, where is that "export options" documented? still relevant for app router?

**Comment by HansLjy** [comment]
Thank you. It seems to be something that I need. But I was wondering how can I disable body parser for all api routes? Because the code you provide looks like a 'middleware' to me. I'd like to use it for every request handler so I may need to disable body parser for all api routes

**Reply by icyJoseph** [reply]
Yeah, I think you ought to do that on an one by one basis... 

Maybe if you create a catch all route inside pages/api? Then technically all routes, unless created separately, will use the same handler. Within that handler you could reject non-valid paths as well, process the body and specific methods, etc. 

https://nextjs.org/docs/api-routes/dynamic-api-routes

A bit of a cop-out though.

**Comment by HansLjy** [comment]
@icyJoseph Sorry to bother you again. I found that after `req.read()`, it became unusable by formidable or busboy. Does it have side effect?

**Reply by icyJoseph** [reply]
Yeah, the body stream can only be read once IIRC, you could move it to the scope where you decode it though. So first check the content type, then process the body within that scope. Specifically needed in this case because those frameworks want to work directly with the stream, however the buffer you collected the body into, is perfectly usable too.
